File: api/app/services/json_to_jsx.rb
require "json"

class JsonToJsx
  def call(tree)
    render_node(tree, 0).rstrip
  end

  private

  def render_node(node, depth)
    element  = node["component"]
    # Separate slot props from regular props
    regular_props = {}
    slot_props = {}

    node.each do |k, v|
      next if k == "component"
      if v.is_a?(Array) && v.any? { |x| x.is_a?(Hash) && x["component"] }
        slot_props[k] = v
      elsif k == "children"
        slot_props["children"] = v
      else
        regular_props[k] = v
      end
    end

    indent = "  " * depth
    open   = "<#{element}#{props_str(element, regular_props)}"

    # Render all slot content (children first, then named slots)
    inner_parts = []
    slot_props.each do |_slot_name, slot_content|
      inner_parts << render_children(slot_content, depth + 1)
    end
    inner = inner_parts.join

    return "#{indent}#{open} />\n" if inner.empty?
    "#{indent}#{open}>\n#{inner}#{indent}</#{element}>\n"
  end

  def render_children(c, depth)
    return "" if c.nil? || c == false || (c.is_a?(String) && c.empty?)

    case c
    when String, Numeric
      text_line(c, depth)
    when Array
      c.map { |x|
        if x.is_a?(Hash) && (x["component"])
          render_node(x, depth)
        else
          text_line(x, depth)
        end
      }.join
    when Hash
      render_node(c, depth)
    else
      text_line(c.to_s, depth)
    end
  end

  def text_line(text, depth)
    "#{"  " * depth}#{text}\n"
  end

  def props_str(element, props)
    parts = serialize_props(element, props)
    parts.empty? ? "" : " " + parts.join(" ")
  end

  def serialize_props(element, props)
    keys = props.keys.map(&:to_s).reject do |k|
      v = props[k] || props[k.to_s]
      v.nil? || (v.is_a?(String) && v.empty?)
    end

    priority = %w[id className variant size disabled style]
    aria = ->(k) { k.start_with?("aria-") || k.start_with?("data-") }

    keys.sort_by! do |k|
      [
        (priority.include?(k) ? 0 : (aria.call(k) ? 1 : 2)),
        k
      ]
    end

    keys.each_with_object([]) do |k, acc|
      v = props[k] || props[k.to_s]

      case v
      when TrueClass, FalseClass
        next if v == false
        acc << "#{k}={true}"
      else
        acc << "#{k}=#{serialize_prop_value(k, v)}"
      end
    end
  end

  def serialize_prop_value(key, val)
    return "{null}" if val.nil?

    case val
    when String
      if val.start_with?("__id:")
        "{#{val.sub('__id:', '')}}"
      else
        "\"#{val.gsub('"', '\"')}\""
      end
    when Numeric
      "{#{val}}"
    when TrueClass, FalseClass
      val ? "{true}" : "{false}"
    when Array, Hash
      if key == "style" && val.is_a?(Hash)
        "{{#{JSON.generate(val)}}}"
      else
        "{#{JSON.generate(val)}}"
      end
    else
      "{#{JSON.generate(val)}}"
    end
  end
end

File: app/src/utils/jsonToJsx.js
/**
 * Client-side port of /api/app/services/json_to_jsx.rb
 * Converts a JSON component tree into JSX string.
 */

const PRIORITY_PROPS = ["id", "className", "variant", "size", "disabled", "style"];

export function jsonToJsx(tree) {
  if (!tree || typeof tree !== "object") return "";
  return renderNode(tree, 0).trimEnd();
}

function renderNode(node, depth) {
  const element = node.component;
  if (!element) return "";

  const regularProps = {};
  const slotProps = {};

  for (const [k, v] of Object.entries(node)) {
    if (k === "component") continue;
    if (Array.isArray(v) && v.some((x) => x && typeof x === "object" && x.component)) {
      slotProps[k] = v;
    } else {
      regularProps[k] = v;
    }
  }

  const indent = "  ".repeat(depth);
  const open = `<${element}${propsStr(element, regularProps)}`;

  const innerParts = [];
  for (const [slotName, slotContent] of Object.entries(slotProps)) {
    const rendered = renderChildren(slotContent, depth + 2);
    if (!rendered) continue;
    const slotIndent = "  ".repeat(depth + 1);
    innerParts.push(`${slotIndent}<Slot name="${slotName}">\n${rendered}${slotIndent}</Slot>\n`);
  }
  const inner = innerParts.join("");

  if (!inner) return `${indent}${open} />\n`;
  return `${indent}${open}>\n${inner}${indent}</${element}>\n`;
}

function renderChildren(c, depth) {
  if (c === null || c === undefined || c === false || c === "") return "";

  if (typeof c === "string" || typeof c === "number") {
    return textLine(c, depth);
  }

  if (Array.isArray(c)) {
    return c
      .map((x) => {
        if (x && typeof x === "object" && x.component) {
          return renderNode(x, depth);
        }
        return textLine(x, depth);
      })
      .join("");
  }

  if (typeof c === "object" && c.component) {
    return renderNode(c, depth);
  }

  return textLine(String(c), depth);
}

function textLine(text, depth) {
  return `${"  ".repeat(depth)}${text}\n`;
}

function propsStr(element, props) {
  const parts = serializeProps(element, props);
  return parts.length ? " " + parts.join(" ") : "";
}

function serializeProps(element, props) {
  const keys = Object.keys(props).filter((k) => {
    const v = props[k];
    return v !== null && v !== undefined && !(typeof v === "string" && v === "");
  });

  keys.sort((a, b) => {
    const rankA = propRank(a);
    const rankB = propRank(b);
    if (rankA !== rankB) return rankA - rankB;
    return a.localeCompare(b);
  });

  const result = [];
  for (const k of keys) {
    const v = props[k];
    if (typeof v === "boolean") {
      if (v === false) continue;
      result.push(`${k}={true}`);
    } else {
      result.push(`${k}=${serializePropValue(k, v)}`);
    }
  }
  return result;
}

function propRank(k) {
  if (PRIORITY_PROPS.includes(k)) return 0;
  if (k.startsWith("aria-") || k.startsWith("data-")) return 1;
  return 2;
}

function serializePropValue(key, val) {
  if (val === null || val === undefined) return "{null}";

  if (typeof val === "string") {
    if (val.startsWith("__id:")) {
      return `{${val.slice(5)}}`;
    }
    if (key.endsWith("Component") && /^[A-Z]/.test(val)) {
      return `{${val}}`;
    }
    if (val.includes('"')) {
      return `{"${val.replace(/\\/g, "\\\\").replace(/"/g, '\\"')}"}`;
    }
    return `"${val}"`;
  }

  if (typeof val === "number") return `{${val}}`;
  if (typeof val === "boolean") return val ? "{true}" : "{false}";

  if (key === "style" && typeof val === "object" && !Array.isArray(val)) {
    return `{{${JSON.stringify(val)}}}`;
  }

  return `{${JSON.stringify(val)}}`;
}
